File: RegexFileSorter/Controls/TextBoxFolder.vb
Imports System.IO
Imports Microsoft.WindowsAPICodePack.Dialogs

Public Class TextBoxFolder
	Inherits TextBox
	Public Shadows ReadOnly IsReadonly As Boolean
	Private Shared ReadOnly COFD As New CommonOpenFileDialog() With {.IsFolderPicker = True}

	Public Sub New()
		MyBase.IsReadOnly = True
	End Sub

	Public Sub SelectFolder()
		COFD.InitialDirectory = Text
		If COFD.ShowDialog = CommonFileDialogResult.Ok Then Text = COFD.FileName
	End Sub

	Protected Overrides Sub OnDragEnter(e As DragEventArgs)
		'MyBase.OnDragEnter(e)
		ProcessDrag(e)
	End Sub

	Protected Overrides Sub OnDragLeave(e As DragEventArgs)
		'MyBase.OnDragLeave(e)
	End Sub

	Protected Overrides Sub OnDragOver(e As DragEventArgs)
		'MyBase.OnDragOver(e)
		ProcessDrag(e)
	End Sub

	Protected Overrides Sub OnDrop(e As DragEventArgs)
		'MyBase.OnDrop(e)
		Dim Folder = DirectCast(e.Data.GetData(DataFormats.FileDrop), String())(0)
		If Directory.Exists(Folder) Then Text = Folder
	End Sub

	Private Sub ProcessDrag(e As DragEventArgs)
		Dim Effect = DragDropEffects.None
		If e.Data.GetDataPresent(DataFormats.FileDrop) Then
			Dim Folder = DirectCast(e.Data.GetData(DataFormats.FileDrop), String())(0)
			If Directory.Exists(Folder) Then
				Effect = DragDropEffects.Link
			End If
		End If
		e.Effects = Effect
		e.Handled = True
	End Sub

End Class

File: RegexFileSorter/Controls/TextBoxFolder.cs
using Microsoft.WindowsAPICodePack.Dialogs;
using System.IO;
using System.Windows;
using System.Windows.Controls;

namespace RegexFileSorter.Controls
{
    internal class TextBoxFolder : TextBox
    {
        public TextBoxFolder()
        {
            base.IsReadOnly = true;
        }

        public new bool IsReadOnly { get; } = true;

        public void SelectFolder()
        {
            var F = new CommonOpenFileDialog
            {
                IsFolderPicker = true,
                InitialDirectory = Text
            };
            if (F.ShowDialog() == CommonFileDialogResult.Ok) { Text = F.FileName; }
        }

        protected override void OnDragEnter(DragEventArgs e) => ProcessDrag(e);

        protected override void OnDragLeave(DragEventArgs e) => base.OnDragLeave(e);

        protected override void OnDragOver(DragEventArgs e) => ProcessDrag(e);

        protected override void OnDrop(DragEventArgs e)
        {
            var Folder = ((string[])e.Data.GetData(DataFormats.FileDrop))[0];
            if (Directory.Exists(Folder)) { Text = Folder; }
        }

        private static void ProcessDrag(DragEventArgs e)
        {
            var Effect = DragDropEffects.None;
            if (e.Data.GetDataPresent(DataFormats.FileDrop))
            {
                var Folder = ((string[])e.Data.GetData(DataFormats.FileDrop))[0];
                if (Directory.Exists(Folder)) { Effect = DragDropEffects.Link; }
            }
            e.Effects = Effect;
            e.Handled = true;
        }
    }
}